tree/dataframe/test/dataframe_datasetspec.py

```diff
@@ -0,0 +1,93 @@
+import unittest
+import ROOT
+import os
+
+RDataFrame = ROOT.ROOT.RDataFrame
+RDatasetSpec = ROOT.RDF.Experimental.RDatasetSpec
+REntryRange = ROOT.RDF.Experimental.RDatasetSpec.REntryRange
+
+class RDatasetSpecTest(unittest.TestCase):
+    @classmethod
+    def setUp(cls):
+        cls.legacy_pyroot = os.environ.get('LEGACY_PYROOT') == 'on'
+
+        # reuse the code from the C++ unit tests to create some files
+        code = ''' {
+        auto dfWriter0 = ROOT::RDataFrame(5)
+                            .Define("x", [](ULong64_t e) { return e; }, {"rdfentry_"})
+                            .Define("y", [](ULong64_t e) { return e + 10; }, {"rdfentry_"})
+                            .Define("z", [](ULong64_t e) { return e + 100; }, {"rdfentry_"});
+        dfWriter0.Range(0, 2).Snapshot<ULong64_t>("subTree", "PYspecTestFile2.root", {"z"});
+        dfWriter0.Range(2, 4).Snapshot<ULong64_t>("subTree", "PYspecTestFile3.root", {"z"});
+        dfWriter0.Range(4, 5).Snapshot<ULong64_t>("subTree", "PYspecTestFile4.root", {"z"});
+        dfWriter0.Range(0, 2).Snapshot<ULong64_t>("subTree1", "PYspecTestFile5.root", {"z"});
+        dfWriter0.Range(2, 4).Snapshot<ULong64_t>("subTree2", "PYspecTestFile6.root", {"z"});
+        dfWriter0.Snapshot<ULong64_t>("anotherTree", "PYspecTestFile7.root", {"z"});
+        }'''
+        ROOT.gInterpreter.Calc(code)
+
+    def tearDown(self):
+        for i in range(2, 8):
+            ROOT.gSystem.Unlink("PYspecTestFile" + str(i) + ".root")
+
+    # a test involving all 3 constructors, all possible valid ranges, all possible friends
+    def test_General(self):
+        ranges = [REntryRange(), REntryRange(1, 4), REntryRange(2,4), REntryRange(100), REntryRange(1, 100),
+                  REntryRange(2, 2), REntryRange(7, 7)]
+        expectedRess = [[100, 101, 102, 103, 104], [101, 102, 103], [102, 103], [100, 101, 102, 103, 104],
+                        [101, 102, 103, 104], [], []]
+        specs = []
+        
+        for r in ranges:
+            specs.append([])
+            specs[-1].append(RDatasetSpec("anotherTree", "PYspecTestFile7.root", r))
+            specs[-1].append(RDatasetSpec("subTree", ["PYspecTestFile2.root",
+                                                 "PYspecTestFile3.root",
+                                                 "PYspecTestFile4.root"], r))
+            specs[-1].append(RDatasetSpec([("subTree1", "PYspecTestFile5.root"),
```

Comment:
So here you are creating an empty list, then appending an empty list as the last element, then taking this inner empty list and appending to it a bunch of `RDatasetSpect` objects, am I right? Why is that?

Reply:
```py
                if self.legacy_pyroot:
                    resP = rdf.Take("ULong64_t")("z")
                    ...
                else:
                    resP = rdf.Take["ULong64_t"]("z")
```

Reply:
right. I want to test all ctors and at the same time make it rather "organized"